ddl/ddl_worker_test.go

```diff
@@ -1060,6 +1068,60 @@ func (s *testDDLSuite) TestCancelJob(c *C) {
 	c.Assert(len(baseTable.Meta().Partition.Definitions), Equals, 2)
 	c.Assert(baseTable.Meta().Partition.Definitions[1].ID, Equals, addedPartInfo.Definitions[0].ID)
 	c.Assert(baseTable.Meta().Partition.Definitions[1].LessThan[0], Equals, addedPartInfo.Definitions[0].LessThan[0])
+
+	// Cancel modify column which should reorg the data.
+	c.Assert(failpoint.Enable("github.com/pingcap/tidb/ddl/skipMockContextDoExec", `return(true)`), IsNil)
+	baseTableInfo = testTableInfoWith2IndexOnFirstColumn(c, d, "modify-table", 2)
+	// This will cost 2 global id, one for table id, the other for the job id.
+	testCreateTable(c, ctx, d, dbInfo, baseTableInfo)
+
+	cancelState = model.StateNone
+	newCol := baseTableInfo.Columns[0].Clone()
+	// change type from long to tinyint.
+	newCol.FieldType = *types.NewFieldType(mysql.TypeTiny)
+	// change from null to not null
+	newCol.FieldType.Flag |= mysql.NotNullFlag
+	newCol.FieldType.Flen = 2
+
+	originColName := baseTableInfo.Columns[0].Name
+	pos := &ast.ColumnPosition{Tp: ast.ColumnPositionNone}
+
+	updateTest(&tests[49])
+	modifyColumnArgs = []interface{}{&newCol, originColName, pos, mysql.TypeNull, 0}
+	doDDLJobErrWithSchemaState(ctx, d, c, dbInfo.ID, baseTableInfo.ID, test.act, modifyColumnArgs, &cancelState)
+	c.Check(checkErr, IsNil)
+	baseTable = testGetTable(c, d, dbInfo.ID, baseTableInfo.ID)
+	c.Assert(baseTable.Meta().Columns[0].FieldType.Tp, Equals, mysql.TypeLong)
+	c.Assert(baseTable.Meta().Columns[0].FieldType.Flag&mysql.NotNullFlag, Equals, uint(0))
```

Comment:
If it‘s rollbacked, why this flag is `NotNullFlag`.

Reply:
the result is asserted equal to 0, which means there is no `NotNullFlag`
it is the same as `c.Assert(baseTable.Meta().Columns[0].FieldType.Flag&mysql.NullFlag, Equals, uint(1))`

Reply:
I've changed to `c.Assert(mysql.HasNotNullFlag(baseTable.Meta().Columns[0].FieldType.Flag), Equals, false)`